Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ProblemsController.cs

```diff
@@ -372,59 +381,73 @@ public ActionResult Delete(int? id)
             }
 
             if (!this.CheckIfUserHasProblemPermissions(id.Value))
             {
                 this.TempData.AddDangerMessage(GlobalConstants.NoPrivilegesMessage);
-                return this.RedirectToAction("Index", "Contests", new { area = "Administration" });
+                return this.RedirectToAction<ContestsController>(c => c.Index());
             }
 
-            var selectedProblem = this.Data.Problems.All()
-                .Where(x => x.Id == id)
-                .Select(DetailedProblemViewModel.FromProblem)
+            var selectedProblem = this.problemsData
+                .GetByIdQuery(id.Value)
+                .Select(DeleteProblemViewModel.FromProblem)
                 .FirstOrDefault();
 
             if (selectedProblem == null)
             {
                 this.TempData.AddDangerMessage(GlobalResource.Invalid_problem);
-                return this.RedirectToAction(nameof(this.Index));
+                return this.RedirectToAction(c => c.Index());
+            }
+
+            if (this.contestsData.CanBeCompetedById(selectedProblem.ContestId))
+            {
+                this.TempData.AddDangerMessage(GlobalResource.Active_contest_problems_permitted_for_deletion);
+                return this.RedirectToAction(c => c.Contest(selectedProblem.ContestId));
             }
 
             return this.View(selectedProblem);
         }
 
-        public ActionResult ConfirmDelete(int? id)
+        [HttpPost]
+        [ValidateAntiForgeryToken]
+        public ActionResult ConfirmDelete(int problemId)
         {
-            if (id == null)
+            if (!this.CheckIfUserHasProblemPermissions(problemId))
             {
-                this.TempData.AddDangerMessage(GlobalResource.Invalid_problem);
-                return this.RedirectToAction(nameof(this.Index));
+                this.TempData.AddDangerMessage(GlobalConstants.NoPrivilegesMessage);
+                return this.RedirectToAction<ContestsController>(c => c.Index());
             }
 
-            var problem = this.Data.Problems.All().FirstOrDefault(x => x.Id == id);
+            var problem = this.problemsData.GetWithContestById(problemId);
 
             if (problem == null)
             {
                 this.TempData.AddDangerMessage(GlobalResource.Invalid_problem);
-                return this.RedirectToAction(nameof(this.Index));
+                return this.RedirectToAction(c => c.Index());
             }
 
-            this.Data.Resources.Delete(r => r.ProblemId == id);
+            if (problem.Contest.CanBeCompeted)
+            {
+                this.TempData.AddDangerMessage(GlobalResource.Active_contest_problems_permitted_for_deletion);
+                return this.RedirectToAction(c => c.Contest(problem.ContestId));
+            }
 
-            this.Data.TestRuns.Delete(tr => tr.Submission.ProblemId == id);
+            this.Data.Resources.Delete(r => r.ProblemId == problemId);
```

Comment:
Why call the `Resources` repository instead of the `Problems` one?

Reply:
I call the problemsDataService for the deletion of the problem